src/coreclr/jit/optimizer.cpp

```diff
@@ -4575,6 +4575,267 @@ PhaseStatus Compiler::optUnrollLoops()
+                        (op1->gtFlags & GTF_SIDE_EFFECT) != 0 || (op2->gtFlags & GTF_SIDE_EFFECT) != 0)
+                    {
+                        return false;
+                    }
+                    asg_node = tree;
+                    asg_stmt = stmt;
+
+                    if (op2->gtOper == GT_SELECT)
+                    {
+                        found_select = true;
+                    }
+                    break;
+                }
+
+                // These do not need conditional execution.
+                case GT_NOP:
+                    break;
+
+                // Cannot optimise this block.
+                default:
+                    return false;
+            }
+        }
+    }
+    if (asg_node == nullptr)
+    {
+        // The blocks checked didn't contain any ASG nodes.
+        return false;
+    }
+
+#ifdef DEBUG
+    if (verbose)
+    {
+        JITDUMP("Attempting to conditionally execute " FMT_BB " from " FMT_BB "\n", middle_block->bbNum, block->bbNum);
+        for (BasicBlock* dump_block = block; dump_block != middle_block->bbNext; dump_block = dump_block->bbNext)
+        {
+            fgDumpBlock(dump_block);
+        }
+        JITDUMP("\n");
+    }
+#endif
+
+    if (found_select)
+    {
+        // The assign is already conditional. Try adding another condition.
+        if (!optIfConvertCCmp(last->gtGetOp1(), asg_node))
+        {
+            return false;
+        }
+    }
+    else
+    {
+        // Make the assign conditional.
+        if (!optIfConvertSelect(last->gtGetOp1(), asg_node))
+        {
+            return false;
+        }
+    }
+
+    // Remove the JTRUE statement.
+    last->ReplaceWith(gtNewNothingNode(), this);
+    fgSetStmtSeq(block->lastStmt());
+    gtSetEvalOrder(last);
+    fgSetStmtSeq(asg_stmt);
+
+    // Update the flow.
+    fgRemoveAllRefPreds(block->bbJumpDest, block);
+    block->bbJumpKind = BBJ_NONE;
+    block->bbJumpDest = middle_block;
+
+#ifdef DEBUG
+    if (verbose)
+    {
+        JITDUMP("After if conversion\n");
+        for (BasicBlock* dump_block = block; dump_block != middle_block->bbNext; dump_block = dump_block->bbNext)
+        {
+            fgDumpBlock(dump_block);
+        }
+        JITDUMP("\n");
+    }
+#endif
+
+    return true;
+#endif
+}
+
+bool Compiler::optIfConvertSelect(GenTree* original_condition, GenTree* asg_node)
+{
+    assert(original_condition->OperIsCmpCompare());
+    assert(asg_node->gtOper == GT_ASG);
+
+    // Duplicate the input of the assign.
+    // This will be used as the false result of the select node.
+    GenTree* current_value = gtCloneExpr(asg_node->AsOp()->gtOp1);
+    current_value->gtFlags &= GTF_EMPTY;
+
+    // Duplicate the condition and invert it
+    GenTree* cond = gtCloneExpr(original_condition);
+    cond->gtFlags |= GTF_DONT_CSE;
+    cond->gtFlags ^= GTF_RELOP_JMP_USED;
+    cond->gtFlags ^= GTF_RELOP_NAN_UN;
+    cond->gtOper = GenTree::ReverseRelop(cond->gtOper);
```

Comment:
Seems like it would be better to flip `op1` and `op2` instead, in particular since the condition would not necessarily have to be a compare in the future. Does that somehow result in worse codegen?

Reply:
`op1 < op2` flipped is `op2 < op1`, which isn't the same as reversing: `op1 < op2` becomes `op1 >= op2`

Also, op2 could be an immediate.

Reply:
I mean flip `op1` and `op2` of the `SELECT` node (created in the next line), not of the comparison.

Reply:
Ah, right, of course :)

With the code as it is:
```
if (op1 > 3 && op2 != 10) {
    op1 = 5;
}
```
becomes:
```
IN0003: 000010                    cmp    w1, #10
IN0004: 000014                    ccmp    w0, #3, nzc, ne
IN0005: 000018                    csel    w0, w3, w0, gt
```
Having the conditions in the assembly match the code makes reading it much easier.

However,  thinking about it now, ideally the `ne 10` compare should be first, then the `gt 3` compare. That's something to fix elsewhere though